site/src/pages/tutorials/thrift/blog/implement-read.mdx

```diff
@@ -0,0 +1,250 @@
+---
+menuTitle: "Implement READ"
+order: 5
+type: step
+category: thrift
+tags:
+  - server
+level: basic
+---
+
+# Implementing READ operation
+
+In the earlier step, we created blog posts.
+In this step, we'll implement a read operation and make a call to read blog posts.
+We'll write two service methods, one for reading a single post and another for multiple posts.
+
+<TutorialSteps current={5} />
+
+## What you need
+
+You need to have the following obtained from previous steps:
+
+- [Generated Java code](/tutorials/thrift/blog/create-service#2-compile-the-thrift-file)
+- `Main.java`
+- `BlogServiceImpl.java`
+- `BlogClient.java`
+- `BlogServiceTest.java`
+
+## 1. Implement server-side
+
+Let's write two methods for retrieving blog posts; one for a single post and another for multiple posts.
+
+<Tabs>
+<TabPane tab="Single post" key="1">
+
+1. Implement the `getBlogPost()` method in the `BlogServiceImpl` class to retrieve a single post.
+  ```java filename=BlogServiceImpl.java
+  @Override
+  public void getBlogPost(GetBlogPostRequest request, AsyncMethodCallback<BlogPost> resultHandler)
+          throws TException {
+    final BlogPost blogPost = blogPosts.get(request.getId());
+    resultHandler.onComplete(blogPost);
+  }
+  ```
+2. Add an if-else statement to throw an exception in case there is no blog post for the given ID.
+  ```java filename=BlogServiceImpl.java
+  import example.armeria.blog.thrift.BlogNotFoundException;
+  ...
+  @Override
+  public void getBlogPost(GetBlogPostRequest request, AsyncMethodCallback<BlogPost> resultHandler)
+          throws TException {
+    final BlogPost blogPost = blogPosts.get(request.getId());
+    if (blogPost == null) {
+      // throwing an exception will also have the same effect
+      // throw new BlogNotFoundException("The blog post does not exist. ID: " + request.getId());
+      resultHandler.onError(
+              new BlogNotFoundException("The blog post does not exist. ID: " + request.getId()));
+    } else {
+      resultHandler.onComplete(blogPost);
+    }
+  }
+  ```
```

Comment:
Optional) I'm not strong on this, but I'm wondering if it's necessary to divide this into two parts. (I guess users will end up just copy-pasting the bottom version anyways)

```suggestion
Implement the `getBlogPost()` method in the `BlogServiceImpl` class to retrieve a single post.
  ```java filename=BlogServiceImpl.java
  import example.armeria.blog.thrift.BlogNotFoundException;
  ...
  @Override
  public void getBlogPost(GetBlogPostRequest request, AsyncMethodCallback<BlogPost> resultHandler)
          throws TException {
    final BlogPost blogPost = blogPosts.get(request.getId());
    if (blogPost == null) {
      // throwing an exception will also have the same effect
      // throw new BlogNotFoundException("The blog post does not exist. ID: " + request.getId());
      resultHandler.onError(
              new BlogNotFoundException("The blog post does not exist. ID: " + request.getId()));
    } else {
      resultHandler.onComplete(blogPost);
    }
  }
  ```
```